polar-core/src/kb.rs

```diff
@@ -1,26 +1,42 @@
 use std::collections::HashMap;
 
-use crate::error::PolarResult;
+use crate::error::ParameterError;
+use crate::error::{PolarError, PolarResult};
 
 pub use super::bindings::Bindings;
 use super::counter::Counter;
 use super::rules::*;
 use super::sources::*;
 use super::terms::*;
+use std::sync::Arc;
 
-/// A map of bindings: variable name → value. The VM uses a stack internally,
-/// but can translate to and from this type.
+enum RuleParamMatch {
+    True,
+    False(String),
+}
```

Comment:
Feel like this would be better as

```suggestion
type RuleParamMismatch = Option<String>;
```

so we can leverage all of the Rust combinators over `Option`, but don't feel too strongly about it

Reply:
Don't really want to do this refactor at this point but could do it as a follow up